src/common/cockpitcloserange.c

```diff
@@ -0,0 +1,92 @@
+#include "cockpitcloserange.h"
+
+#include <assert.h>
+#include <dirent.h>
+#include <err.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/resource.h>
+
+static int
+closefd (int from, int max_fd, int fd)
+{
+  if (fd >= from && fd <= max_fd)
+    {
+      while (close (fd) < 0)
+        {
+          if (errno == EAGAIN || errno == EINTR)
+            continue;
+          if (errno == EBADF || errno == EINVAL)
+            break;
+          warnx ("couldn't close fd: %m");
+          return -1;
+        }
+    }
+
+  return 0;
+}
+
+int
+cockpit_close_range (int from, int max_fd, int flags)
+{
+  int open_max;
+  int fd;
+  int res = 0;
+
+  struct rlimit rl;
+
+  /* we want to keep the API compatible to glibc's upcoming close_range(), but don't implement flags */
+  assert (flags == 0);
+
+#ifdef __linux__
+  DIR *d;
+
+  if ((d = opendir ("/proc/self/fd"))) {
+      struct dirent *de;
+
+      while ((de = readdir (d))) {
+          long l;
+          char *e = NULL;
+
+          if (de->d_name[0] == '.')
+              continue;
+
+          errno = 0;
+          l = strtol (de->d_name, &e, 10);
+          if (errno != 0 || !e || *e)
+              continue;
+
+          fd = (int) l;
+
+          if ((long) fd != l)
+              continue;
+
+          if (fd == dirfd (d))
+              continue;
+
+          if ((res = closefd (from, max_fd, fd)) != 0)
```

Comment:
Just put the condition here and call `close()` directly?

All of the error checking above is a bit weird imho.  `close()` is always successful. Any error you get there is delayed from a previously failed operation (and the fd is still closed).  The obvious exception is EBADF, but even in that case you end up with a closed fd.

Reply:
Done, thanks for pointing out that close() is always successful in Linux -- that helps indeed!